src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs

```diff
@@ -0,0 +1,250 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class NullableWalker
+    {
+        /// <summary>
+        /// Maintains a cache of the most recent variable states and manages
+        /// taking a snapshot of the NullableWalker in order to be as efficient
+        /// as possible when creating many snapshots.
+        /// </summary>
+        private sealed class Snapshotter
+        {
+            private readonly NullableWalker _walkerToSnapshot;
+            private ImmutableDictionary<VariableIdentifier, int> _lastVariableSlot = ImmutableDictionary<VariableIdentifier, int>.Empty;
+            private ImmutableArray<VariableIdentifier> _lastVariableBySlot = ImmutableArray<VariableIdentifier>.Empty;
+            private readonly PooledDictionary<BoundNode, Snapshot> _snapshotMap;
+            private readonly bool _ownsMap;
+
+            internal Snapshotter(NullableWalker walker, bool takeIncrementalSnapshot)
+            {
+                _walkerToSnapshot = walker;
+                if (takeIncrementalSnapshot)
+                {
+                    _snapshotMap = PooledDictionary<BoundNode, Snapshot>.GetInstance();
```

Comment:
>_snapshotMap = PooledDictionary<BoundNode, Snapshot>.GetInstance(); [](start = 20, length = 67)

It feels like it would be better to have this map be always supplied by a consumer. Consumer owns the map and is responsible for freeing it and turn to immutable when appropriate.

Reply:
This way sharing the same map across different Snapshotter instances will be more natural and straightforward.

---
In reply to: [288789402](https://github.com/dotnet/roslyn/pull/35850#discussion_r288789402) [](ancestors = 288789402)

Reply:
I'm moving this to a more standard builder model, and I don't think this comment is relevant anymore.

---
In reply to: [288798291](https://github.com/dotnet/roslyn/pull/35850#discussion_r288798291) [](ancestors = 288798291,288789402)